### PR TITLE
per-bucket random replica order #2957

### PR DESCRIPTION
**Pull request by yaojingguo**
[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3041)

### CODE REVIEW

**Comment by tamird** [comment]
@yaojingguo you don't need to close the PR and open a new one; if you update your branch, your PR will be updated.

**Comment by yaojingguo** [comment]
@tamird, got it, sorry for the trouble.

**Comment by yaojingguo** [comment]
There is a race failure TestReplicateRogueRemovedNode in CircleCI. But I have successfully run `./build/circle-local.sh` on my laptop. How can I proceed from here?

**Comment by tamird** [comment]
I rebuilt it for you, green now.

**Comment by lizhongz** [comment]
Hi @yaojingguo , I have also worked on this issue. Your implementation is good except two small pieces.
* The maxIndex parameter of randPerm method is not necessary.
* The rand.Perm method is *pseudo-random*, which means the results will be the same every time you call it with the same argument. As a result, you should set a seed before calling it.

**Comment by yaojingguo** [comment]
@lizhongz 

> The maxIndex parameter of randPerm method is not necessary.

I have originally thought that `len`'s cost is O(n). After I read [Go Slices: usage and internals](https://blog.golang.org/go-slices-usage-and-internals), I know that `len`'s cost is O(1).  So I have removed `maxIndex` according to your suggestion.

> The rand.Perm method is pseudo-random, which means the results will be the same every time you call it with the same argument. As a result, you should set a seed before calling it.

`pseudo-random` means that the sequence is generated for the same seed.  It does not mean that `the results will be the same every time you call it with the same argument`. The default rand source has been initialized with `rand.Seed(randutil.NewPseudoSeed())` in ` main.go`.  So I think that we don't need to do it again here.

**Comment by lizhongz** [comment]
@yaojingguo Good to know that the seed has been set in main.go.

**Comment by tschottdorf** [comment]
Could you write some tests? For testing `randPerm` it will be better if you pass the random number generator in as a parameter (and not use `rand.*` directly).

File: kv/replica_slice.go [hunk 1]
```diff
@@ -103,7 +105,8 @@ func (rs replicaSlice) SortByCommonAttributePrefix(attrs []string) int {
 		return 0
 	}
 	topIndex := len(rs) - 1
-	for bucket := 0; bucket < len(attrs); bucket++ {
+	bucket := 0
```
Comment (tschottdorf) [inline kv/replica_slice.go]: could you back out this part of the change (doesn't seem like an improvement) or explain why you prefer it this way?
Also, remove the TODO on line 102.

**Comment by yaojingguo** [comment]
@tschottdorf, I have created a new commit according to your suggestions.

File: kv/replica_slice_test.go [hunk 2]
```diff
@@ -110,3 +117,25 @@ func TestReplicaSetMoveToFront(t *testing.T) {
 		t.Errorf("expected order %s, got %s", exp, stores)
 	}
 }
+
+func TestReplicaSetRandPerm(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	rs := createReplicaSlice()
+	r := rand.New(rand.NewSource(randutil.NewPseudoSeed()))
```
Comment (tamird) [inline kv/replica_slice_test.go]: don't you want a known seed for this test? `r := rand.New(rand.NewSource(0))`?

File: kv/replica_slice_test.go [hunk 3]
```diff
@@ -110,3 +117,25 @@ func TestReplicaSetMoveToFront(t *testing.T) {
...
+	}
+	rs.randPerm(2, 2, r.Perm)
+	if stores := getStores(rs); !reflect.DeepEqual(stores, exp) {
+		t.Errorf("expected order %s, got %s", exp, stores)
+	}
+
+	rs.randPerm(0, 2, r.Perm)
+	stores := getStores(rs)[3:5]
+	exp = exp[3:5]
+	if !reflect.DeepEqual(stores, exp) {
```
Comment (tschottdorf) [inline kv/replica_slice_test.go]: as @tamird suggested, get use a known seed and verify that you actually get rearranged replica sets (in particular guarding against one-offs: get an example in which the first and last item of the shuffled part of the slice are not the same).

**Comment by yaojingguo** [comment]
@tamird and @tschottdorf, a new commit according to your suggestions. Now the slice shuffle allocates a new slice because it make the verification of the shuffled slice ordering easy. If allocating a new slice is disallowed, I can do the shuffle in place.

**Comment by tamird** [comment]
There is no need to allocate a new slice in `randPerm`. If you need a copy, make the copy in your test.

**Comment by yaojingguo** [comment]
@tamird, a new commit which does  `in-place shuffle` and avoids the allocation of a new slice.

File: kv/replica_slice_test.go [hunk 4]
```diff
@@ -110,3 +116,32 @@ func TestReplicaSetMoveToFront(t *testing.T) {
 		t.Errorf("expected order %s, got %s", exp, stores)
 	}
 }
+
+func TestReplicaSetRandPerm(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	rs := createReplicaSlice()
+	r := rand.New(rand.NewSource(0))
+	rs.randPerm(0, 4, r.Intn)
+	exp := []roachpb.StoreID{1, 2, 3, 4, 5}
```
Comment (tschottdorf) [inline kv/replica_slice_test.go]: can you add a test case in which `5` moves?

**Comment by tschottdorf** [comment]
LGTM once the test case is added. The test failure likely requires a rebase.

**Comment by yaojingguo** [comment]
@tschottdorf , a new commit after:
 
1. A rebase is performed.
2. Add more tests as you have suggested and make some other minor changes. 

For details, check the commit message and the code.

**Comment by tschottdorf** [comment]
Thanks @yaojingguo! Please squash everything (one easy way to do that is `git fetch origin && git reset --soft origin/master && git commit`) and then we can merge it.

**Pull request closed without merging**